### PR TITLE
Shooter callbacks

### PR DESCRIPTION
**Pull request by cannedlock**

### CODE REVIEW

File: components/breakSensors.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+from enum import IntEnum
+from components.shooterLogic import ShooterLogic
+# from components.ledLogic import LED
+
+class Sensors(IntEnum):
+    kLoadingSensor = 5
+    kShootingSensor = 1
+
+class BreakSensors:
+
```
Comment (tuffinmuffin) [inline components/breakSensors.py]: Doc strings missing

File: components/breakSensors.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    # led: LED
+    logger: logging
+
+    def setup(self):
+        self.entries = {}
+        # self.sensorArray = []
+
+        self.logger.info("IR Break Sensor component created.")
+
+    # def canBeAutomatic(self):
```
Comment (tuffinmuffin) [inline components/breakSensors.py]: Future code or dead code? Comme nt if you want to use for later work

File: components/breakSensors.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        entry = {}
+        entry["stateMachine"] = stateMachine
+        if "Loader" in actionType:
+            entry["actionType"] = "Loader"
+        elif "Shooter" in actionType:
+            entry["actionType"] = "Shooter"
+        elif "LED" in actionType:
+            entry["actionType"] = "LED"
+        else:
+            raise ValueError
```
Comment (tuffinmuffin) [inline components/breakSensors.py]: Give ValueError a text string about the error ValueError("Invalid Action Type")

File: components/breakSensors.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        sensor - what sensor is used to execute this action (MUST be an int)
+        requestedValue - what value the sensor has to read to execute this action
+        conditionalState - state the StateMachine must be in for callback to execute (None if no state necessary)
+        newState - what state the shooterLogic StateMachine transitions to
+        """
+        self.logger.info("sensor %s, requesting %s, condition %s, moving to %s.", int(sensor), requestedValue, conditionalState, newState)
+        entry = {}
+        entry["stateMachine"] = stateMachine
+        if "Loader" in actionType:
+            entry["actionType"] = "Loader"
```
Comment (tuffinmuffin) [inline components/breakSensors.py]: This can be a good enum use.

File: configs/doof.yml [hunk 5]
```diff
@@ -64,6 +63,22 @@ loaderMotors:
       inverted: False
       pid: Null
 
+loaderSensors:
+  subsystem: "breaksensors"
+  description: "All IR break sensors used for the loader live here for the loader/LED component"
+  loaderSensors:
+    groups: "sensors"
+    sensor1:
```
Comment (tuffinmuffin) [inline configs/doof.yml]: You can give them names here if you wanted. sensor1 loadingSensor, sensor5 topElevator maybe? Or at least a Description: "info on sensor purpose" field

File: utils/sensorFactories.py [hunk 6]
```diff
@@ -24,5 +25,25 @@ def gyroFactory(descp):
...
         logging.error("Failed to create gyro for %s. Error %s",descp, e)
-    
     return None
+
+def breaksensorFactory(descp):
+    """
+    Creates break sensors from a break sensor descp
+    """
+    try:
+        if 1 == descp["channel"]:
```
Comment (tuffinmuffin) [inline utils/sensorFactories.py]: Should use the value channel as dio(dscp["channel"])

File: components/breakSensors.py (hunk 1 above)
Comment (cannedlock) [reply components/breakSensors.py]: added basic docstrings

File: components/breakSensors.py (hunk 2 above)
Comment (cannedlock) [reply components/breakSensors.py]: future code, added a note

File: components/breakSensors.py (hunk 3 above)
Comment (cannedlock) [reply components/breakSensors.py]: added string

File: components/breakSensors.py (hunk 4 above)
Comment (cannedlock) [reply components/breakSensors.py]: added enum

File: utils/sensorFactories.py (hunk 6 above)
Comment (cannedlock) [reply utils/sensorFactories.py]: fixed

File: configs/doof.yml (hunk 5 above)
Comment (cannedlock) [reply configs/doof.yml]: added descriptions

File: components/breakSensors.py (hunk 1 above)
Comment (ChrisHirsch) [reply components/breakSensors.py]: Please resolve once resolved

File: components/shooterLogic.py [hunk 7]
```diff
@@ -94,58 +84,58 @@ def runLoaderManually(self):
...
     @state
-    def runShooter(self, state_tm):
+    def shootRunShooter(self, state_tm):
         """Runs shooter to a certain speed or until a set time."""
+        self.shooterMotors.stopLoader()
         self.shooterMotors.runShooter(1)
-        if self.shooterMotors.shooterMotor.getEncoder().getVelocity() >= self.targetShootingSpeed or state_tm > 2:
-            self.next_state('shoot')
+        if self.shooterMotors.shooterMotor.getEncoder().getVelocity() >= self.targetShootingSpeed or state_tm > 3:
+            self.next_state('shootFire')
```
Comment (ChrisHirsch) [inline components/shooterLogic.py]: I like this! This means we are going to compete in Robot Wars now right?

**Pull request merged**